rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,554 @@
+the status endpoint.
+
+We may also want to have the `/status` endpoint display each individual worker status in addition to the 
+global status, which may be breaking change in the `/status` API response format.
+
+### PID file
+
+Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 
+usage, as the PID stored in the file will be a arbitrary worker’s PID, instead of the coordinator (main process) PID.
+
+In clustering mode, we will need to have to coordinator handle the PID file logic, and to disable pid file handling 
+in the worker's environment service.
+
+### SavedObjects migration
+
+In the current state, all workers are going to try to perform the migration. Ideally, we would have only one process 
+perform the migration, and the other ones just wait for a ready signal. We can’t easily have the coordinator do it, 
+so we would probably have to leverage the ‘main worker’ concept here.
+
+The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
+Kibana, so this can probably be considered an improvement.
+
+## Open questions and things to solve
+
+### Memory consumption
+
+In clustered mode, node options such as `max-old-space-size` will be used by all processes. 
+
+E.g using `--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 
+
+If this something we will need to document somewhere?
+
+### Workers error handling
+
+When using `cluster`, the common best practice is to have the coordinator recreate ('restart') workers when they terminate unexpectedly. 
+However, given Kibana's architecture, some failures are not recoverable (workers failing because of config validation, failed migration...). 
+
+For instance, if a worker (well, all workers) terminates because of an invalid configuration property, it doesn't make
+any sense to have the coordinator recreate them indefinitely, as the error requires manual intervention. 
+
+Should we try to distinguish recoverable and non-recoverable errors, or are we good terminating the main Kibana process 
+when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?
+
+### Data folder
+
+The data folder (`path.data`) is currently the same for all workers. 
+
+We still have to identify with the teams if this is going to be a problem. It could be, for example, if some plugins
+are accessing files in write mode, which could result in concurrency issues between the workers.
+
+If that was confirmed, we would have to create and use a distinct data folder for each worker.
+
+One pragmatic solution could be, when clustering is enabled, to create a sub folder under path.data for each worker. 
+
+The data folder is not considered part of our public API, and the implementation and path already changed in previous
```

Comment:
Really? How did it change in previous minor releases?

Reply:
This statement is coming from @joshdover, I'll let him answer if he remembers

Reply:
I'm not sure what I said and I'm having a hard time finding an instance where we changed this, I must have misspoke!

However, I don't believe we consider the _structure_ of the data folder to be part of our public API. We should be able to use the sub-folder approach if needed I believe.

Reply:
What all is stored in the data folder? Is it just the `uuid` file? The docs are not really clear...

>
path.data: | The path where Kibana stores persistent data not saved in Elasticsearch. Default: data
-- | --

Source: https://www.elastic.co/guide/en/kibana/current/settings.html

And then they're contradicted by the package directory layout
> 

data | The location of the data files written to disk by Kibana and its plugins | /var/lib/kibana
-- | -- | --
path.data | logs | Logs files location | /var/log/kibana
path.logs | plugins | Plugin files location. Each plugin will be contained in a subdirectory. | /usr/share/kibana/plugins

Source: https://www.elastic.co/guide/en/kibana/current/rpm.html#rpm-layout

Reply:
The only plugin that I have found that is referencing this is the reporting plugin, however I can't find a place in code the actually reads the configuration. @elastic/kibana-reporting-services Could you provide any guidance into how the `path.data` config is used by the Reporting plugin? My guess is as some sort of temporary storage, however I can't find any usages.

Reply:
Pinging @elastic/kibana-app-services for any guidance on what reporting does here.

Reply:
I think we used to use it for the "user data directory" for Chromium, but as I look in the code now, we just use `tmpdir` from the `os` package: https://github.com/elastic/kibana/blob/master/x-pack/plugins/reporting/server/browsers/chromium/driver_factory/index.ts#L51